🐛  (kustomize/v1 and kustomize/v2-alpha) : ComponentConfig scaffolds should not be done by default

laxmikantbpandhare opened the pull request:
## A description of the change

This PR will verify whether the `--component-config` flag is passed or not. If not then it will not scaffold out the `controller_manager_config.yaml` file.

## The motivation for the change
it fixes #2782  

## Tests done to ensure the changes: 

Init command ( without the --component-config flag)

From this PR:  
```yaml
laxmikantbhaskarpandhare@lpandhar-mac test % kubebuilder init --domain my.domain --repo my.domain/guestbook

Writing kustomize manifests for you to edit...
false
Writing scaffold for you to edit...
Get controller runtime:
$ go get sigs.k8s.io/controller-runtime@v0.12.2
Update dependencies:
$ go mod tidy
Next: define a resource with:
$ kubebuilder create api
laxmikantbhaskarpandhare@lpandhar-mac test % tree
.
├── Dockerfile
├── Makefile
├── PROJECT
├── README.md
├── config
│   ├── default
│   │   ├── kustomization.yaml
│   │   ├── manager_auth_proxy_patch.yaml
│   │   └── manager_config_patch.yaml
│   ├── manager
│   │   ├── kustomization.yaml
│   │   └── manager.yaml
│   ├── prometheus
│   │   ├── kustomization.yaml
│   │   └── monitor.yaml
│   └── rbac
│       ├── auth_proxy_client_clusterrole.yaml
│       ├── auth_proxy_role.yaml
│       ├── auth_proxy_role_binding.yaml
│       ├── auth_proxy_service.yaml
│       ├── kustomization.yaml
│       ├── leader_election_role.yaml
│       ├── leader_election_role_binding.yaml
│       ├── role_binding.yaml
│       └── service_account.yaml
├── go.mod
├── go.sum
├── hack
│   └── boilerplate.go.txt
└── main.go

6 directories, 24 files
```

Init command ( with the changes and passed the --component-config flag)

```yaml
laxmikantbhaskarpandhare@lpandhar-mac testData % kubebuilder init --domain my.domain --repo my.domain/guestbook --component-config

Writing kustomize manifests for you to edit...
true
Writing scaffold for you to edit...
Get controller runtime:
$ go get sigs.k8s.io/controller-runtime@v0.12.2
Update dependencies:
$ go mod tidy
Next: define a resource with:
$ kubebuilder create api
laxmikantbhaskarpandhare@lpandhar-mac testData % tree
.
├── Dockerfile
├── Makefile
├── PROJECT
├── README.md
├── config
│   ├── default
│   │   ├── kustomization.yaml
│   │   ├── manager_auth_proxy_patch.yaml
│   │   └── manager_config_patch.yaml
│   ├── manager
│   │   ├── controller_manager_config.yaml
│   │   ├── kustomization.yaml
│   │   └── manager.yaml
│   ├── prometheus
│   │   ├── kustomization.yaml
│   │   └── monitor.yaml
│   └── rbac
│       ├── auth_proxy_client_clusterrole.yaml
│       ├── auth_proxy_role.yaml
│       ├── auth_proxy_role_binding.yaml
│       ├── auth_proxy_service.yaml
│       ├── kustomization.yaml
│       ├── leader_election_role.yaml
│       ├── leader_election_role_binding.yaml
│       ├── role_binding.yaml
│       └── service_account.yaml
├── go.mod
├── go.sum
├── hack
│   └── boilerplate.go.txt
└── main.go

6 directories, 25 files
```

This will verify that it is working, the `controller_manager_config.yaml` file got scaffolded only when a flag(`--component-config`) is passed.

laxmikantbpandhare commented:
@camilamacedo86 - I verified the field `--component-config` and then scaffolded accordingly. I have a quick question, I did these changes in `v1`. Is there an expectation to do it in `v2-alpha` as well?

camilamacedo86 commented:
Hi @laxmikantbpandhare,

> @camilamacedo86 - I verified the field --component-config and then scaffolded accordingly. I have a quick question, I did these changes in v1. Is there an expectation to do it in v2-alpha as well?

Terrific ! yes, we need to also fix the v2-alpha kustomize plugin.

camilamacedo86 commented:
/ok-to-test

laxmikantbpandhare commented:
@camilamacedo86 - I wen through the tests that are failing and found that they are failing in `kustomize build`

It is because `kustomization.yaml` file as shown below, it required `controller_manager_config.yaml` file to be present. But as we did the change, that only allow that file to be generated when `component-config` got passed.

```
resources:
- manager.yaml

generatorOptions:
  disableNameSuffixHash: true

configMapGenerator:
- name: manager-config
  files:
  - controller_manager_config.yaml
```


The solution for this is, that I need to modify the test cases and enable the `component-config` field so that the `controller_manager_config.yaml` file will be generated WDYT?

camilamacedo86 commented on pkg/plugins/common/kustomize/v1/scaffolds/init.go:
```diff
@@ -62,23 +62,44 @@ func (s *initScaffolder) Scaffold() error {
-		&prometheus.Kustomization{},
-		&prometheus.Monitor{},
-	)
+	if s.config.IsComponentConfig() {
```
@laxmikantbpandhare, 

Could we just add here https://book.kubebuilder.io/component-config-tutorial/api-changes.html the step to add this manifest manually when the project was not created with --component-config?

laxmikantbpandhare replied:
working on it.

laxmikantbpandhare replied:
Updated the docs.

camilamacedo86 commented on test/e2e/deployimage/plugin_cluster_test.go:
```diff
@@ -94,6 +94,7 @@ var _ = Describe("kubebuilder", func() {
 				"--plugins", "go/v3",
 				"--project-version", "3",
 				"--domain", kbc.Domain,
+				"--component-config=true",
```
@laxmikantbpandhare 
Why we are adding it here?

laxmikantbpandhare replied:
@camilamacedo86 - As explained [here](https://github.com/kubernetes-sigs/kubebuilder/pull/2826#issuecomment-1184974419), It is failing with `kustomize build`

laxmikantbpandhare replied:
I re-ran after removal, you can see tests are failing.

camilamacedo86 commented on test/e2e/deployimage/plugin_cluster_test.go:
```diff
@@ -139,6 +140,7 @@ var _ = Describe("kubebuilder", func() {
 				"--plugins", "go/v3",
 				"--project-version", "3",
 				"--domain", kbc.Domain,
+				"--component-config=true",
```
Same ?

camilamacedo86 commented on test/e2e/v4/generate_test.go:
```diff
@@ -43,6 +43,7 @@ func GenerateV4(kbc *utils.TestContext) {
 		"--plugins", "go/v4-alpha",
 		"--project-version", "3",
 		"--domain", kbc.Domain,
+		"--component-config=true",
```
Same could remove it or clear a mock test only for this?

laxmikantbpandhare replied:
If I remove this flag, the test will fail because the` Kustomization.yaml` file is expecting the manager component config file to be generated.

laxmikantbpandhare commented:
/label tide/merge-method-squash

camilamacedo86 commented:
Hi @laxmikantbpandhare, 

Regards the comment : https://github.com/kubernetes-sigs/kubebuilder/pull/2826#issuecomment-1184974419

That means: 
- We should ONLY add to the default kustomization the content which is relevant to the component config when the flag is used (So we also need to address this one)
- We also need to add this step to the documents for those that want to use the component-config on projects that were not init with the flag 
- We need to have a **new** test on the e2e tests for v3 and v4 that verifies a project scaffold with component-config 

Could you look for to address these needs?

laxmikantbpandhare commented:
> Hi @laxmikantbpandhare,
> 
> Regards the comment : [#2826 (comment)](https://github.com/kubernetes-sigs/kubebuilder/pull/2826#issuecomment-1184974419)
> 
> That means:
> 
> * We should ONLY add to the default kustomization the content which is relevant to the component config when the flag is used (So we also need to address this one)
> * We also need to add this step to the documents for those that want to use the component-config on projects that were not init with the flag
> * We need to have a **new** test on the e2e tests for v3 and v4 that verifies a project scaffold with component-config
> 
> Could you look to address these needs? WDYT?

@camilamacedo86 - Got your point and agree with the changes. I will update the PR soon.

SayakMukhopadhyay commented:
Hi @camilamacedo86 I am trying to align my own project to incorporate these changes manually (since its already scaffolded). A clarification regarding the last comment which might be important for this PR too
> We should ONLY add to the default kustomization the content which is relevant to the component config when the flag is used

Is the `generatorOptions` required when creating the project without the config flag. In other words, without the flag should the `kustomization.yaml` be
```yaml
resources:
- manager.yaml
```
or
```yaml
resources:
- manager.yaml

generatorOptions:
  disableNameSuffixHash: true
```
I am pretty sure that all of 
```yaml
configMapGenerator:
- name: manager-config
  files:
  - controller_manager_config.yaml
```
needs to be removed for sure.

SayakMukhopadhyay commented on pkg/plugins/common/kustomize/v1/scaffolds/init.go:
```diff
@@ -62,23 +62,44 @@ func (s *initScaffolder) Scaffold() error {
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
```
```suggestion
			&kdefault.ManagerAuthProxyPatch{},
			&prometheus.Kustomization{},
```

laxmikantbpandhare replied:
As per the discussion, I will be making the patch conditional based. If the user specifies, the `--component-config` flag then only `kustomization.yml` file will have that patch. I am working on changes.

camilamacedo86 replied:
@laxmikantbpandhare 

The `&kdefault.ManagerConfigPatch{},` is used in the default scaffold?
I mean, have we any reason to do the scaffold if/when we are not looking the use of the ComponentConfig?

If not, then we do not need to scaffold so, we need to check it.

laxmikantbpandhare replied:
@camilamacedo86 @SayakMukhopadhyay - the scaffold of `kustomization.yaml` from `/config/default` location scaffolds some of the part depending upon `componenet-config`. We are planning to use the same login in `/config/manager` too. I will update the PR, and every question will be answered.

laxmikantbpandhare replied:
Code changes completed. Working on docs and testing part.

laxmikantbpandhare replied:
There is no need to change in `/config/default`. It is because the `kustomization.yml` file commented out the file `manager_config_patch.yaml` whenever there is no flag passed. So, this file only gets triggered if the flag is passed through the command line.

SayakMukhopadhyay replied:
Yeah, its true that the `kustomization.yaml` has the patch line commented out as I noted in this [comment](https://github.com/kubernetes-sigs/kubebuilder/pull/2826#pullrequestreview-1046430684). The reason I would prefer this file to be conditionally scaffolded too as the presence of this file could confuse first time users as the file was not _strictly_ getting used anywhere.

Yeah, its mentioned in the commented line in `kustomization.yaml` but so is `manager_webhook_patch.yaml` and `webhookcainjection_patch.yaml` but those files are not created unless needed.

laxmikantbpandhare replied:
Agree with you, if the `--component-config` field is not passed then these two files are getting created unnecessarily.
Let's see what @camilamacedo86 says on this.

camilamacedo86 replied:
I think it makes sense close this one as sorted out.

SayakMukhopadhyay commented on pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go:
```diff
@@ -62,23 +62,45 @@ func (s *initScaffolder) Scaffold() error {
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
```
```suggestion
			&kdefault.ManagerAuthProxyPatch{},
			&prometheus.Kustomization{},
```

laxmikantbpandhare replied:
Even, if we keep these file as it is, it will not have any problem.

laxmikantbpandhare replied:
Started working on it.

SayakMukhopadhyay reviewed (changes requested):
I dont know if this is something that you intend to fix but just thought to mention that the `ManagerConfigPatch` is also something that should depend on whether the config flag is pased or not. The generated patch is referencing the `controller_manager_config.yaml` which won't exist and is sure to confuse users.
https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/kdefault/manager_config_patch.go

Also notice how the `kustomize.yaml` referencing the above patch is commenting out the line to include the patch

https://github.com/kubernetes-sigs/kubebuilder/blob/e116b5423ff00d1f9a2075039a29916e1d173319/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/kdefault/kustomization.go#L79-L84

laxmikantbpandhare commented:
> Hi @camilamacedo86 I am trying to align my own project to incorporate these changes manually (since its already scaffolded). A clarification regarding the last comment which might be important for this PR too
> 
> > We should ONLY add to the default kustomization the content which is relevant to the component config when the flag is used
> 
> Is the `generatorOptions` required when creating the project without the config flag. In other words, without the flag should the `kustomization.yaml` be
> 
> ```yaml
> resources:
> - manager.yaml
> ```
> 
> or
> 
> ```yaml
> resources:
> - manager.yaml
> 
> generatorOptions:
>   disableNameSuffixHash: true
> ```
> 
> I am pretty sure that all of
> 
> ```yaml
> configMapGenerator:
> - name: manager-config
>   files:
>   - controller_manager_config.yaml
> ```
> 
> needs to be removed though.

Yes, you are correct.

```
> configMapGenerator:
> - name: manager-config
>   files:
>   - controller_manager_config.yaml
> ```

> ```yaml
> configMapGenerator:
> - name: manager-config
>   files:
>   - controller_manager_config.yaml
> ```

Yes, you are correct.

We are removing below completely.

```
configMapGenerator:
- name: manager-config
  files:
  - controller_manager_config.yaml
  ```

camilamacedo86 commented on test/e2e/v3/generate_test.go:
```diff
@@ -139,6 +139,7 @@ func GenerateV3(kbc *utils.TestContext, crdAndWebhookVersion string) {
 		"--project-version", "3",
 		"--domain", kbc.Domain,
 		"--fetch-deps=false",
+		"--component-config=true",
```
I think we need a new test only for the componentConfig so that we can ensure this scenario

laxmikantbpandhare replied:
yes, will remove this too.

laxmikantbpandhare replied:
Removed.

camilamacedo86 commented on pkg/plugins/common/kustomize/v1/scaffolds/init.go:
```diff
@@ -62,23 +62,44 @@ func (s *initScaffolder) Scaffold() error {
 		machinery.WithConfig(s.config),
 	)
 
```
@laxmikantbpandhare @SayakMukhopadhyay 

It shows that we can also remove 

```
generatorOptions:
  disableNameSuffixHash: true
```

when the flag is not informed because it shows to be used only in the generate option for configComponent

laxmikantbpandhare replied:
Yes, Modified the` kustomization.yml` file for both `v1` and `v2-alpha`.

laxmikantbpandhare commented:
Added code changes. I am working on docs and e2e testing part.

laxmikantbpandhare commented:
> I dont know if this is something that you intend to fix but just thought to mention that the `ManagerConfigPatch` is also something that should depend on whether the config flag is pased or not. The generated patch is referencing the `controller_manager_config.yaml` which won't exist and is sure to confuse users. https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/kdefault/manager_config_patch.go
> 
> Also notice how the `kustomize.yaml` referencing the above patch is commenting out the line to include the patch
> 
> https://github.com/kubernetes-sigs/kubebuilder/blob/e116b5423ff00d1f9a2075039a29916e1d173319/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/kdefault/kustomization.go#L79-L84

@camilamacedo86 - We can apply the same logic mentioned here. Scaffold out file and don't use it in `kustomization.yml`
Both way it work. WDYT?

laxmikantbpandhare commented:
Added e2e test for v3 and v4.

laxmikantbpandhare commented:
> Hi @laxmikantbpandhare,
> 
> Regards the comment : [#2826 (comment)](https://github.com/kubernetes-sigs/kubebuilder/pull/2826#issuecomment-1184974419)
> 
> That means:
> 
> * We should ONLY add to the default kustomization the content which is relevant to the component config when the flag is used (So we also need to address this one)
> * We also need to add this step to the documents for those that want to use the component-config on projects that were not init with the flag
> * We need to have a **new** test on the e2e tests for v3 and v4 that verifies a project scaffold with component-config
> 
> Could you look to address these needs? WDYT?

@camilamacedo86 - Addressed all the review comments.

laxmikantbpandhare commented:
/assign @pwittroc

camilamacedo86 commented on docs/book/src/component-config-tutorial/api-changes.md:
```diff
@@ -67,3 +67,30 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
 
 [tutorial-source]: https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/component-config-tutorial/testdata/project
+
+## If you miss flag `--component-config` during the `kubebuilder init` command, then add it manually as shown below.
```
@laxmikantbpandhare see that we have an subsection `Setting up an exising project` where add the steps about what you need to do if you do not scaffold the project with the flag:
https://github.com/kubernetes-sigs/kubebuilder/blob/9dcd389aa9e2386af20b08e045d0334e788c768b/docs/book/src/component-config-tutorial/api-changes.md#setting-up-an-exising-project

So, ihmo it needs to be +1 step under this subsection instead of a new one.

camilamacedo86 replied:
Also, we need to add here the step to add to the file: 

```
generatorOptions:
  disableNameSuffixHash: true

configMapGenerator:
- name: manager-config
  files:
  - controller_manager_config.yaml
```

laxmikantbpandhare replied:
Updated file and made above as subsection in `Setting up an existing project`. Also, I added the above missing content.

camilamacedo86 reviewed (commented):
@laxmikantbpandhare GREAT WORK 🥇 
All shows fine. Just a nit for the docs, see: https://github.com/kubernetes-sigs/kubebuilder/pull/2826/files#r931596735

laxmikantbpandhare commented:
> @laxmikantbpandhare GREAT WORK 🥇 All shows fine. Just a nit for the docs, see: https://github.com/kubernetes-sigs/kubebuilder/pull/2826/files#r931596735

thank you @camilamacedo86. Updated the document.

laxmikantbpandhare commented:
/retest pull-kubebuilder-e2e-k8s-1-20-7

laxmikantbpandhare commented:
/retest-required

laxmikantbpandhare commented:
/test pull-kubebuilder-e2e-k8s-1-20-7

laxmikantbpandhare commented:
Started working on the `/config/default` flow. I will update the PR soon.

laxmikantbpandhare commented:
/hold need to add one more feature

laxmikantbpandhare commented:
/hold cancel

laxmikantbpandhare commented:
@SayakMukhopadhyay @camilamacedo86 - Updated the PR with the new changes that we discussed for `default/kustomization.ysml` scaffolding.

camilamacedo86 commented on docs/book/src/component-config-tutorial/api-changes.md:
```diff
@@ -67,3 +67,41 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
 
 [tutorial-source]: https://github.com/kubernetes-sigs/kubebuilder/tree/master/docs/book/src/component-config-tutorial/testdata/project
+
+Add below content to this file ->`/config/manager/controller_manager_config.yaml`
```
```suggestion
Create the file "/config/manager/controller_manager_config.yaml" with the following content:
```

laxmikantbpandhare replied:
Updated.

SayakMukhopadhyay commented:
@laxmikantbpandhare please correct me if I am wrong, from what I understand, if the ComponentConfig is disabled
instead of
```yaml
# Mount the controller config file for loading manager configurations
# through a ComponentConfig type
#- manager_config_patch.yaml
```
the resultant section of the `kustomization,yaml` would be
```yaml
# Mount the controller config file for loading manager configurations
# through a ComponentConfig type
```

camilamacedo86 commented on docs/book/src/component-config-tutorial/api-changes.md:
````diff
@@ -67,3 +67,41 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
+# leaderElectionReleaseOnCancel: true
+```
+
+Add below content to this file ->`/config/manager/kustomization.yaml`
````
```suggestion
Update the file `/config/manager/kustomization.yaml` by adding at the botton the following content:
```

laxmikantbpandhare replied:
Updated.

camilamacedo86 reviewed (commented):
@laxmikantbpandhare very nice work.
Just a few nits and could you please squash the commits?

laxmikantbpandhare commented:
> @laxmikantbpandhare please correct me if I am wrong, from what I understand, if the ComponentConfig is disabled instead of
> 
> ```yaml
> # Mount the controller config file for loading manager configurations
> # through a ComponentConfig type
> #- manager_config_patch.yaml
> ```
> 
> the resultant section of the `kustomization.yaml` would be
> 
> ```yaml
> # Mount the controller config file for loading manager configurations
> # through a ComponentConfig type
> ```

@SayakMukhopadhyay Yes, I will remove those comments as well.

laxmikantbpandhare commented:
> @laxmikantbpandhare very nice work. Just a few nits and could you please squash the commits?

Sure @camilamacedo86

camilamacedo86 commented on docs/book/src/component-config-tutorial/api-changes.md:
````diff
@@ -67,3 +67,41 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
+- name: manager-config
+  files:
+  - controller_manager_config.yaml
+```
````
```suggestion
```

Update the `default/kustomization.yaml` by adding under the `patchesStrategicMerge:` the following patch:

```yaml

# Mount the controller config file for loading manager configurations
# through a ComponentConfig type
- manager_config_patch.yaml
```

```

```

camilamacedo86 commented on docs/book/src/component-config-tutorial/api-changes.md:
````diff
@@ -67,3 +67,41 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
+- name: manager-config
+  files:
+  - controller_manager_config.yaml
+```
````
```suggestion
```

Update the `default/kustomization.yaml` by adding under the `patchesStrategicMerge:` the following patch:

```yaml

# Mount the controller config file for loading manager configurations
# through a ComponentConfig type
- manager_config_patch.yaml
```

```

laxmikantbpandhare replied:
Yes, I updated the docs section.

camilamacedo86 reviewed (approved):
Great 🥇 

/approved

In a follow up we need to create a plugin that does this config.
Then, we might also implement the edit subcommand to do the required changes

laxmikantbpandhare commented:
/test pull-kubebuilder-e2e-k8s-1-19-16

everettraven commented on docs/book/src/component-config-tutorial/api-changes.md:
````diff
@@ -67,3 +67,68 @@ mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), options)
+# leaderElectionReleaseOnCancel: true
+```
+
+Update the file `/config/manager/kustomization.yaml` by adding at the botton the following content:
````
Typo:
```suggestion
Update the file `/config/manager/kustomization.yaml` by adding at the bottom the following content:
```

laxmikantbpandhare replied:
Updated.

everettraven commented on pkg/plugins/common/kustomize/v1/scaffolds/init.go:
```diff
@@ -62,23 +62,44 @@ func (s *initScaffolder) Scaffold() error {
+	if s.config.IsComponentConfig() {
+		return scaffold.Execute(
+			&rbac.Kustomization{},
+			&rbac.AuthProxyRole{},
+			&rbac.AuthProxyRoleBinding{},
+			&rbac.AuthProxyService{},
+			&rbac.AuthProxyClientRole{},
+			&rbac.RoleBinding{},
+			&rbac.LeaderElectionRole{},
+			&rbac.LeaderElectionRoleBinding{},
+			&rbac.ServiceAccount{},
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&manager.ControllerManagerConfig{},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
+			&prometheus.Monitor{},
+		)
+	} else {
+		return scaffold.Execute(
+			&rbac.Kustomization{},
+			&rbac.AuthProxyRole{},
+			&rbac.AuthProxyRoleBinding{},
+			&rbac.AuthProxyService{},
+			&rbac.AuthProxyClientRole{},
+			&rbac.RoleBinding{},
+			&rbac.LeaderElectionRole{},
+			&rbac.LeaderElectionRoleBinding{},
+			&rbac.ServiceAccount{},
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
+			&prometheus.Monitor{},
+		)
+	}
```
From what I can tell here, the only difference in the scaffolds that needs to happen is that the template `&manager.ControllerManagerConfig{}` isn't included by default.

To reduce the repetition and increase readability I think we could do something like:

```go
templates := []machinery.Template{
  &rbac.Kustomization{},
  &rbac.AuthProxyRole{},
  &rbac.AuthProxyRoleBinding{},
  &rbac.AuthProxyService{},
  &rbac.AuthProxyClientRole{},
  &rbac.RoleBinding{},
  &rbac.LeaderElectionRole{},
  &rbac.LeaderElectionRoleBinding{},
  &rbac.ServiceAccount{},
  &manager.Kustomization{},
  &manager.Config{Image: imageName},
  &kdefault.Kustomization{},
  &kdefault.ManagerAuthProxyPatch{},
  &kdefault.ManagerConfigPatch{},
  &prometheus.Kustomization{},
  &prometheus.Monitor{},
}

if s.config.IsComponentConfig() {
  templates = append(templates, &manager.ControllerManagerConfig{})
}

return scaffold.Execute(templates...)
```
@camilamacedo86 @laxmikantbpandhare WDYT?

laxmikantbpandhare replied:
Yes, sounds perfect!

laxmikantbpandhare replied:
```
templates := []machinery.Builder{
  &rbac.Kustomization{},
  &rbac.AuthProxyRole{},
  &rbac.AuthProxyRoleBinding{},
  &rbac.AuthProxyService{},
  &rbac.AuthProxyClientRole{},
  &rbac.RoleBinding{},
  &rbac.LeaderElectionRole{},
  &rbac.LeaderElectionRoleBinding{},
  &rbac.ServiceAccount{},
  &manager.Kustomization{},
  &manager.Config{Image: imageName},
  &kdefault.Kustomization{},
  &kdefault.ManagerAuthProxyPatch{},
  &kdefault.ManagerConfigPatch{},
  &prometheus.Kustomization{},
  &prometheus.Monitor{},
}

if s.config.IsComponentConfig() {
  templates = append(templates, &manager.ControllerManagerConfig{})
}

return scaffold.Execute(templates...)
```

laxmikantbpandhare replied:
Updated and ready for review @everettraven

everettraven commented on pkg/plugins/common/kustomize/v2-alpha/scaffolds/init.go:
```diff
@@ -62,23 +62,45 @@ func (s *initScaffolder) Scaffold() error {
+		return scaffold.Execute(
+			&rbac.Kustomization{},
+			&rbac.AuthProxyRole{},
+			&rbac.AuthProxyRoleBinding{},
+			&rbac.AuthProxyService{},
+			&rbac.AuthProxyClientRole{},
+			&rbac.RoleBinding{},
+			&rbac.LeaderElectionRole{},
+			&rbac.LeaderElectionRoleBinding{},
+			&rbac.ServiceAccount{},
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&manager.ControllerManagerConfig{},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
+			&prometheus.Monitor{},
+		)
+	} else {
+		return scaffold.Execute(
+			&rbac.Kustomization{},
+			&rbac.AuthProxyRole{},
+			&rbac.AuthProxyRoleBinding{},
+			&rbac.AuthProxyService{},
+			&rbac.AuthProxyClientRole{},
+			&rbac.RoleBinding{},
+			&rbac.LeaderElectionRole{},
+			&rbac.LeaderElectionRoleBinding{},
+			&rbac.ServiceAccount{},
+			&manager.Kustomization{},
+			&manager.Config{Image: imageName},
+			&kdefault.Kustomization{},
+			&kdefault.ManagerAuthProxyPatch{},
+			&kdefault.ManagerConfigPatch{},
+			&prometheus.Kustomization{},
+			&prometheus.Monitor{},
+		)
+
+	}
```
I would do the same thing here in v2-alpha as the comment in the v1 init.go change.

everettraven replied:
Link to comment in v1: https://github.com/kubernetes-sigs/kubebuilder/pull/2826#discussion_r939060615

laxmikantbpandhare replied:
Agree and Modified accordingly.

everettraven reviewed (changes requested):
Just a couple minor nits from me. Other than that it looks great! Great work @laxmikantbpandhare !

everettraven reviewed (approved):
/lgtm

everettraven commented:
Oops, I forgot that an LGTM on this repo auto merges.

@camilamacedo86 is it okay that this has merged?

laxmikantbpandhare commented:
> Oops, I forgot that an LGTM on this repo auto merges.
> 
> 
> 
> @camilamacedo86 is it okay that this has merged?

No worries, thanks for your comments.
